livekit/src/room/metrics/mod.rs

```diff
@@ -0,0 +1,382 @@
+use tokio_util::sync::CancellationToken;
+
+use super::id::{ParticipantIdentity, TrackSid};
+use super::DataPacketKind;
+
+type GenericMetricLabel<T> = (MetricLabel, T);
+
+const QUALITY_LIMITATIONS: [GenericMetricLabel<&str>; 3] = [
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationCpu, "cpu"),
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationBandwidth, "bandwidth"),
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationOther, "other"),
+];
+
+pub struct RtcMetricsManager {}
+
+impl RtcMetricsManager {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn micro_to_milli(micros: i64) -> i64 {
+        micros / 1000
+    }
+
+    fn get_or_create_index(&self, strings: &mut Vec<String>, string: String) -> u32 {
+        let position: Option<usize> = strings.iter().position(|s| s == &string);
+
+        match position {
+            Some(index) => return (index as u32) + (MetricLabel::PredefinedMaxValue as u32),
+            None => {
+                strings.push(string);
+                return ((strings.len() - 1) as u32) + (MetricLabel::PredefinedMaxValue as u32);
+            }
+        }
+    }
+
+    fn create_metric_sample(&self, timestamp_ms: i64, value: f32) -> MetricSample {
+        MetricSample { timestamp_ms, value, ..Default::default() }
+    }
+
+    fn create_time_series(
+        &self,
+        label: MetricLabel,
+        strings: &mut Vec<String>,
+        samples: Vec<MetricSample>,
+        identity: Option<&ParticipantIdentity>,
+        track_sid: Option<&str>,
+        rid: Option<&str>,
+    ) -> TimeSeriesMetric {
+        let mut time_series = TimeSeriesMetric {
+            label: label as u32,
+            samples,
+            ..Default::default()
+        };
+
+        if let Some(id) = identity {
+            time_series.participant_identity = self.get_or_create_index(strings, id.to_string());
+        }
+
+        if let Some(sid) = track_sid {
+            time_series.track_sid = self.get_or_create_index(strings, sid.to_string());
+        }
+
+        if let Some(r) = rid {
+            time_series.rid = self.get_or_create_index(strings, r.to_string());
+        }
+
+        time_series
+    }
+
+    fn create_time_series_for_metric(
+        &self,
+        strings: &mut Vec<String>,
+        label: MetricLabel,
+        stat_value: f32,
+        stat_timestamp_us: i64,
+        track_sid: String,
+        rid: Option<&str>,
+        identity: Option<&ParticipantIdentity>,
+    ) -> TimeSeriesMetric {
+        let timestamp_ms = Self::micro_to_milli(stat_timestamp_us);
+        let sample = self.create_metric_sample(timestamp_ms, stat_value);
+
+        self.create_time_series(label, strings, vec![sample], identity, Some(&track_sid), rid)
+    }
+
+    pub(crate) async fn collect_metrics(
+        &self,
+        room: &Room,
+        rtc_engine: &RtcEngine,
+        cancellation_token: CancellationToken,
```

Comment:
The cancellation can be done beforehand directly in the task inside Room using  tokio::select!

This function should be cancel safe: 
https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety